Multiversion: get stricter with propagating and filtering extension versions

dvoytenko opened the pull request:
Partial for #33001.

jridgewell commented on src/service/extensions-impl.js:
```diff
@@ -578,16 +580,22 @@ export class Extensions {
+        this.win,
+        extensionId,
+        version,
+        /* latest */ false
```
Shouldn't we be requiring `latest` to be passed in?

dvoytenko replied:
The problem is that `latest` is not known until the extension is actually loaded. There are only two APIs here that are called at that point: `registerExtension` and `reloadExtension`. All other APIs, including `preloadExtension`, `installExtensionForDoc`, etc - a caller doesn't actually know whether a particular version is latest or not.

jridgewell commented on src/service/extension-script.js:
```diff
@@ -159,25 +160,45 @@ export function createExtensionScript(win, extensionId, version) {
+export function getExtensionScripts(
+  win,
+  extensionId,
+  version,
+  latest,
```
I don't think we really  need `version` or `latest` here for any of the uses, and it just causes more confusion. Eg, `isExtensionScriptRequired_`, which won't properly catch `latest` version.

dvoytenko replied:
This case is specifically important for `reloadExtension`. What happens is this: the extension script arrives and we see that it's from a wrong RTV, so we re-request it. We need to do two things: (a) mark the old scripts as "skipped", and (b) insert a script for the fresh RTV.  The (b) is easy since we now have the version, so we just construct the script as `/rtv/${RTV/${ext}-${version}.js`.

(a) is harder however. We don't know how the extension was loaded. It could be done either with `${ext}-${version}.js` or `${ext}-latest.js`. So we need to find either option. Or both. Previously we've updated every single script for the same extension. But I don't think that was right, since some of them could have been of different versions. E.g. a fact that a wrong RTV of `amp-mustache-0.1` has been loaded, doesn't mean that the `amp-mustache-0.2` was also loaded with a wrong RTV.

From what I can tell, this is the only use case that needs to be aware of the `latest` flag. However, all other use cases need at least `version`. E.g. if a parent page has `amp-mustache:0.1` and FIE is asking for `amp-mustache:0.2`, we need to be able to check whether the `0.2` version specifically is already on the page.

jridgewell commented on src/service/extensions-impl.js:
```diff
@@ -263,18 +259,23 @@ export class Extensions {
+      version,
+      latest
+    );
     let scriptElement = scriptsInHead.length > 0 ? scriptsInHead[0] : null;
```
Oh, just noticed this bug. `scriptsInHead` can be an array of `[js, mjs]` scripts, and we don't check for which we actually care about based on the Runtime's ESM setting.

dvoytenko replied:
But does it matter in this context? Or can the `.js`/`.mjs` scripts get out of sync?

jridgewell replied:
On an ESM supporting browser, the `.js` version will never load. And vice-versa. I think we have to install the listener on both, else we could get a never ending promise.

dvoytenko replied:
Will be fixed separately.

jridgewell reviewed (approved):
I'm worried we're going to get multiple evaluations of different version scripts (or, even the same version if `0.2` and `latest` are both included/injected). Can we confirm we prevent the loaded script from executing if a another binary already loaded?

dvoytenko commented:
> I'm worried we're going to get multiple evaluations of different version scripts (or, even the same version if `0.2` and `latest` are both included/injected). Can we confirm we prevent the loaded script from executing if a another binary already loaded?

@jridgewell Good idea. I went ahead and added an explicit check + bailout and I added tests for that situation.
